scripts/coverage_report_jenkins.py

```diff
@@ -374,106 +382,217 @@ def sonar_scan(ssh, project_name, project_version, scanner_url, scanner_dir, ser
     run_sonar_scanner(ssh, scanner_dir, timeout)
 
 
-def main(appliance, jenkins_url, jenkins_user, jenkins_token, job_name):
-    if not jenkins_user or not jenkins_token:
-        try:
-            from cfme.utils import conf
-            jenkins_user = conf.credentials.jenkins_app.user
-            jenkins_token = conf.credentials.jenkins_app.token
-        except (AttributeError, KeyError):
-            raise ValueError(
-                '--jenkins-user and --jenkins-token not provided and credentials yaml does not '
-                'contain the jenkins_app entry with user and token')
-    appliance_version = str(appliance.version).strip()
-    logger.info('Looking for appliance version %s in %s', appliance_version, job_name)
-    client = jenkins.Jenkins(jenkins_url, username=jenkins_user, password=jenkins_token)
-    build_numbers = get_build_numbers(client, job_name)
+def get_eligible_builds(jenkins_data, jenkins_job, cfme_version):
+    """Get eligible builds for a specified jenkins job
+
+    An eligible build will be for the specified appliance version, and contain
+    the code coverage data.  We return these builds as a list of named tuples
+    with the following keys: number, job, coverage_archive.
+
+    Args:
+        jenkins_data:   Named tupple with these attributes:  url, user, token, client
+        jenkins_job:  Jenkins job name such as downstream-59z-tests
+        cfme_version:  Versuion CFME sources this coverage is against.
+
+    Returns:
+        List of eligible builds.  Each build is a named tuple with the following
+        keys:  number, job, coverage_archive.
+    """
+    logger.info('Looking for CFME version %s in %s', cfme_version, jenkins_job)
+    build_numbers = get_build_numbers(jenkins_data.client, jenkins_job)
     if not build_numbers:
-        raise Exception('No builds for job {}'.format(job_name))
+        raise Exception('No builds for job {}'.format(jenkins_job))
 
     # Find the builds with appliance version
-    eligible_build_numbers = set()
+    eligible_builds = set()
     for build_number in build_numbers:
+        build = namedtuple('Build', ['number', 'job', 'coverage_archive'])
```

Comment:
that bit is fundamentally broken, it creates a new class each iteration, and assigns class scoped attributes

Reply:
Oh, I can put this outside the for loop correct?

Reply:
BTW, @mshriver suggested using namedtuple, but the location of where I used was my own choice.

Reply:
@jamesooden namedtuple creates a type, not a instance - instances of namedtuple are immutable, thus not suitable for the use in the loop as is

Reply:
@jamesooden you would define the namedtuple class at the module level, and then create an instance of that nametuple here in the loop.

Reply:
@mshriver, gotcha.